### PR TITLE
bpo-47037: Don't test for strftime('%4Y') on Windows

### PR DESCRIPTION
**Pull request by tiran**
https://bugs.python.org/issue47037

### CODE REVIEW

File: Lib/test/support/__init__.py [hunk 1]
```diff
@@ -521,10 +521,13 @@ def requires_subprocess():
...
-    has_strftime_extensions = time.strftime("%4Y") != "%4Y"
-except ValueError:
-    has_strftime_extensions = False
+has_strftime_extensions = False
+if sys.platform != "win32":
+    # bpo-47037: Windows debug builds crash with "Debug Assertion Failed"
+    try:
+        has_strftime_extensions = time.strftime("%4Y") != "%4Y"
+    except ValueError:
+        pass
```
Comment (eryksun) [inline Lib/test/support/__init__.py:530]: The Universal CRT in Windows is developed more aggressively and less conservatively than the old CRT, so it wouldn't surprise me if it eventually supports the "%4Y" format code. I think it should be checked. For example:

```suggestion
try:
    # bpo-47037: Windows debug builds report a failed debug assertion
    with suppress_msvcrt_asserts():
        has_strftime_extensions = time.strftime("%4Y") != "%4Y"
except ValueError:
    has_strftime_extensions = False
```

Comment (eryksun) [reply Lib/test/support/__init__.py:530]: The above suggestion depends on extending the existing `suppress_msvcrt_asserts()` function to return a context manager, and either moving this check down or moving the definition of `suppress_msvcrt_asserts()` up. Here's an example implementation that uses `contextlib.contextmanager`:

```python
def suppress_msvcrt_asserts(verbose=False):
    """Suppress Windows error reporting dialogs."""
    error_mode = crt_report_mode = None

    @contextlib.contextmanager
    def restore_msvcrt_asserts():
        try:
            yield
        finally:
            try:
                if crt_report_mode is not None:
                    for rt, (rm, rf) in crt_report_mode.items():
                        msvcrt.CrtSetReportMode(rt, rm)
                        msvcrt.CrtSetReportFile(rt, rf)
            finally:
                if error_mode is not None:
                    msvcrt.SetErrorMode(error_mode)

    try:
        import msvcrt
    except ImportError:
        return restore_msvcrt_asserts()

    # N.B. SEM_NOALIGNMENTFAULTEXCEPT is not set here. It should be set once
    # for a process, if required. It changes OS behavior, not just error
    # reporting, and it cannot be unset.
    error_mode = msvcrt.SetErrorMode(msvcrt.SEM_FAILCRITICALERRORS
                                       | msvcrt.SEM_NOGPFAULTERRORBOX
                                       | msvcrt.SEM_NOOPENFILEERRORBOX)

    # CrtSetReportMode() is only available in debug builds
    if hasattr(msvcrt, 'CrtSetReportMode'):
        crt_report_mode = {}
        for rt in (msvcrt.CRT_WARN, msvcrt.CRT_ERROR, msvcrt.CRT_ASSERT):
            if verbose:
                rm = msvcrt.CrtSetReportMode(rt, msvcrt.CRTDBG_MODE_FILE)
            else:
                rm = msvcrt.CrtSetReportMode(rt, 0)
            rf = msvcrt.CrtSetReportFile(rt, msvcrt.CRTDBG_FILE_STDERR)
            crt_report_mode[rt] = (rm, rf)

    return restore_msvcrt_asserts()
```

Comment (tim-one) [reply Lib/test/support/__init__.py:530]: What now? @eryksun has "a right way" to fix this, but I don't think @tiran is "a Windows person". If he isn't, he may (understandably!) be uncomfortable with trying to implement such a _relatively_ large change for a platform he doesn't use.

Comment (tiran) [reply Lib/test/support/__init__.py:530]: I don't have access to a Windows system and would appreciate if somebody else takes over the PR. Or I can land my version now to unblock Terry and @eryksun can create a better fix later.

Comment (eryksun) [reply Lib/test/support/__init__.py:530]: Sorry, I'll post the suggestion to extend `suppress_msvcrt_asserts()` to the issue tracker. Are you okay with using the existing `SuppressCrashReport` context manager that's defined later on in this module?

Comment (tim-one) [reply Lib/test/support/__init__.py:530]: @tiran, if I were you I'd commit this, so Terry (& I, among others) can get unstuck ASAP. A principled fix can wait.

**Comment by zooba** [comment]
Another option is to make `has_strftime_extensions` a callable that raises the skip exception if the extensions are not available, then call it inside tests rather than using the decorator. That way it won't be called until everything is set up.

**Comment by zooba** [comment]
I want the expedient fix, so it's going in. We can do something better later, or just wait for someone to complain that the UCRT has added support for this extension and it fails tests in other ways.

**Pull request merged**